src/utils/objectUtils.ts

```diff
@@ -0,0 +1,78 @@
+export function copyWithoutUndefined<T extends Record<K, V>, K extends string | number | symbol, V>(
+  originalValue: T,
+): T {
+  return Object.keys(originalValue).reduce((acc, key) => {
+    // @ts-ignore
+    if (originalValue[key] !== undefined) {
+      // @ts-ignore
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+      acc[key] = originalValue[key]
+    }
+    return acc
+  }, {} as Record<string, unknown>) as T
+}
+
+export function pick<T, K extends string | number | symbol>(
+  source: T,
+  propNames: readonly K[],
+): Pick<T, Exclude<keyof T, Exclude<keyof T, K>>> {
+  const result = {} as T
+  let idx = 0
+  while (idx < propNames.length) {
+    // @ts-ignore
+    if (propNames[idx] in source) {
+      // @ts-ignore
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+      result[propNames[idx]] = source[propNames[idx]]
+    }
+    idx += 1
+  }
+  return result
+}
+
+export function pickWithoutUndefined<T, K extends string | number | symbol>(
+  source: T,
+  propNames: readonly K[],
+): Pick<T, Exclude<keyof T, Exclude<keyof T, K>>> {
+  const result = {} as T
+  let idx = 0
+  while (idx < propNames.length) {
+    // @ts-ignore
+    if (propNames[idx] in source && source[propNames[idx]] !== undefined) {
+      // @ts-ignore
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+      result[propNames[idx]] = source[propNames[idx]]
+    }
+    idx += 1
+  }
+  return result
+}
+
+export function isEmptyObject(params: Record<string, unknown>): boolean {
+  for (const key in params) {
+    if (Object.prototype.hasOwnProperty.call(params, key) && params[key] !== undefined) {
+      return false
+    }
+  }
+  return true
+}
+
+export function groupBy<T>(inputArray: T[], propName: string): Record<string, T[]> {
+  return inputArray.reduce((result, entry) => {
+    // @ts-ignore
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-assignment
+    const key = entry[propName]
+
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+    if (Object.hasOwnProperty.call(result, key)) {
+      // @ts-ignore
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-call
+      result[key].push(entry)
+    } else {
+      // @ts-ignore
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+      result[key] = [entry]
+    }
+    return result
+  }, {})
+}
```

Comment:
🟥

There are bugs hiding in here.

In particular there is no check that the value retrieved by the propName is a valid property key.

Consider the following call:

```ts
const result = groupBy(
    [
        {
            id: {},
        },
    ],
    'id',
);
```

That passes the type check against the current code, but at runtime will try to use an empty object as a key on the result type. JavaScript being JavaScript, it'll try to toString the object before using it as a key, so you'll end up with `result` equal to `{ '[object Object]': { id: {} }}`

Reply:
Here's an alternative implementation:

Note the explicit thrown error inside the reduce body. That's necessary to check the safety of the group value being used.

```ts
export function groupBy<T extends Record<PropertyKey, unknown>, PropName extends keyof T>(
    inputArray: T[],
    propName: PropName,
): Record<string, T[]> {
    return inputArray
        .map((entry) => ({ group: entry[propName], entry }))
        .reduce((acc, { group, entry }) => {
            if (group === null || typeof group !== 'string') {
                throw new Error('foo');
            }
            return { ...acc, [group]: [...(acc[group] ?? []), entry] };
        }, {} as Record<string, T[]>);
}
```

Reply:
Alternatively you might choose to skip bad entries:

```ts
export function groupBy<T extends Record<PropertyKey, unknown>, PropName extends keyof T>(
    inputArray: T[],
    propName: PropName,
): Record<string, T[]> {
    return inputArray
        .map((entry) => ({ group: entry[propName] as unknown, entry }))
        .filter(
            (entry): entry is { group: string; entry: T } => entry.group !== null && typeof entry.group === 'string',
        )
        .reduce(
            (acc, { group, entry }) => ({ ...acc, [group]: [...(acc[group] ?? []), entry] }),
            {} as Record<string, T[]>,
        );
}
```